File: W6/W6D1/make_change.rb
def make_change(target, coins = [25, 10, 5, 1])
  return [] if target == 0
  return nil if coins.none? { |coin| coin < target }
  coins = coins.sort.reverse
  best_change = nil

  coins.each_with_index do |coin, index|
    next if coin > target

    # use this coin
    remainder = target - coin
    best_remainder = make_change(remainder, coins.drop(index))
    next if best_remainder.nil?
    this_change = [coin] + best_remainder

    # Is this better than anything we've seen so far?
    if (best_change.nil? || (this_change.count < best_change.count))
      best_change = this_change
    end
  end

  best_change
end

File: W6/W6D1/make_change.js
var makeChange = function(target, coins) {
  var coins = coins.sort(function(a,b){return(a - b);}).reverse();
  console.log(coins);
  var best_change = null;

  if (target == 0) {
    return [];
  } else if (coins.length == 0) {
    return null;
  } else {
    for(var i = 0; i < coins.length; i++){
      if(target < coins[i]) {
        continue;
      } else {
        var remainder = target - coins[i];
        var best_remainder = makeChange(remainder, coins.slice(i, coins.length));
          if (best_remainder == null) {
            continue;
          }
        var this_change = best_remainder.concat(coins[i]);

        if (best_change == null || this_change.count < best_change.count) {
          best_change = this_change;
        }
      }
    }
  }
  return best_change;
}